### ISSUE TITLE
:terminal improvements

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#5431]
WIP list of terminal improvements.

- User-initiated `:!` invocations should silently use `:terminal` instead (this is purely a UI decision, doesn't affect `:read !foo` etc. See also #5396
- `:term` buffer should be `'modifiable'`.
  - Enables user to manage "scrollback" #2637
- `:term` buffers should `setlocal bufhidde=hide` by default. Fixes #3340 #3138 #5122
- should respect all `'bufhidden'` flags #2368
- should have a ftplugin that sets various window-local options and autocmds
  - "press any key to exit" behavior should be served by a autocmd which is easily disabled
- `:te` should open in split, similar to `:help`
  - support `<mods>` (`:vert`, `:tab`)

**Comment by AprilArcus** [comment]
I think it would be really cool if terminal could support cmdwin - c.f. https://github.com/brettanomyces/nvim-terminus/issues/7

**Comment by Bew78LesellB** [comment]
would it be possible to handle `:set nowrap` in terminal buffer ?

**Comment by justinmk** [comment]
Yes, by telling libvterm a very long width. I have a pending PR that helps with this.

**Comment by mayanksuman** [comment]
For output that spans multiple pages/screen, it will be useful to provide the support of motion keys also.

Right now no motion or scrolling is possible (nvim --version 0.2.0-139g8c9cccb; Debian 8.6; tmux+zsh).

**Comment by justinmk** [comment]
? Please report as a separate bug with steps to reproduce. Motion/scrolling already should work.

**Comment by Bew78LesellB** [comment]
for the following improvment : 
> User-initiated `:!` invocations should silently use `:terminal` instead

Which terminal should be used ? We could send a `RequestTerminal` event, and see if a particular variable is set. If it is, use the terminal from the variable, or create a new terminal (new tab ? h/v split ?)

**Comment by justinmk** [comment]
The details of each item can be discussed as PRs are proposed for them. But at least initially the goal is only to match the traditional semantics of `:!`, so some sort of long-lived scratch buffer would be used.

**Comment by bohrshaw** [comment]
Well, I see (the last line of) a terminal buffer could serve as a really flexible readline-like interface, *if only* we can enter Insert Mode in a terminal buffer.

**Comment by justinmk** [comment]
@bohrshaw See this thread: https://github.com/neovim/neovim/pull/6142#issuecomment-282690785

**Comment by nelstrom** [comment]
Most of the items in the checklist have a PR of their own. Currently, there's no PR associated with this item:

- `:te` should open in split, similar to `:help`

Could we create a PR for that (unless it already exists) and link to it from that item in the checklist? It would be good to have a designated place to discuss that item.

**Comment by tssm** [comment]
I know that I may loose the argument here, but I think Neovim `:terminal` behaves the right way already, and I would whish `:help` and other "special" buffers (e.g. quick fix) behave like it. Here are the reasons:

1. Vim tabs are actually the right design, each tab is a layout and the user can switch between them easily. When `:help` (and `:copen` and probably others that I don't know about) is run the whole point of having different layouts on each tab is lost.
2. Most commands behave the same, with exceptions, so the user must remember those especial cases to avoid disrupting the layout. Even if something like `:0te` is implemented (I would prefer something like `:current te` if it happens), there would be a default behaviour and a bunch of special cases :( I prefer consistency.

One can argue that the user that the user could have custom commands with the desire behaviour, and that's what I do, but it would be awesome if Neovim where more consistent than Vim, so I wouldn't need to remember to use my custom commands instead of the default `:help`. I don't expect the behaviour of `:help` to be changed, I know compatibility with Vim is more important, but it would be nice if `:terminal` was kept as is (plus support for `:vertical`, `:tab` and others)

**Comment by jcorbin** [comment]
Tossing my hat/plugin/angle into the ring, I wrote a small-ish plugin for how I like to use `:terminal`:
- https://github.com/jcorbin/neovim-termhide
- basically I live the "no more tmux, neovim is my root terminal app" lifestyle
- and on top of that, I mostly only over use one, session-specific, terminal at a time
- ... that I want to be able to effectively show/hide, but never really terminate
- sometimes I want it in a "quake-style" top-N line pull down form ( using say `winfixheight` )
- sometimes I want it in a vert-split when doing more involved things
- either way, the missing atom that most of my plugin is oriented around, is "create or show existing" , and make sure that it doesn't get killed when hidden
- ...very much in line with tmux (and screen before it)'s idiom of "create or attach to prior", but now pivoted internally towards "the one terminal"
- uncommonly, one can also spawn alternate shells or REPLs as in `:Term ipython` or `:Term node`

One of my current hanging problems tho, is that I very much wish that I could exclude my terminal buffers from session saving and/or have that `bufhidden` setting restored ( https://github.com/jcorbin/neovim-termhide/issues/2 )

**Comment by ddickstein** [comment]
@justinmk can you elaborate on the plan (or lack thereof) for `:read`? In the original post you write, "`:!` should use `:terminal` internally. Doesn't affect `:read !foo` etc." and then say "`:read :term foo` should work like `:read !foo`" What does this mean? It sounds like `:read !foo` isn't going to be changed by any of the `:!` plans, and that `:read :term foo` is going to be a new supported syntax that happens to be identical to the current `:read !foo`. Is that correct?

**Comment by tmillr** [comment]
> terminal windows in a session are not properly restored

It'd be cool if there were the option to have scrollback restored (or an api to be able to do this manually, like get and set scrollback functions).
